### ISSUE TITLE
Working Python Script to rip sleepcasts

### ISSUE THREAD

**Issue by komali2** [Joexv/HeadRipper#11]
In case you found this repo when trying to rip Headspace Sleepcasts, below is a python script that accomplishes this, using the very helpful research performed by joexv that I was unable to do myself because of APK compilation issues. Thanks, joexv!

First, you need to get a long bearer token, which you can still do today following the instructions in this repo.

Then, you need to get a JSON file with the up to date sleepcast info.

```
curl -k -L -s --compressed 'https://api.prod.headspace.com/content-aggregation/v1/content/view-models/library/topics-category-menu?location=SLEEP&tag=&topicId=41' -H 'Accept: */*' -H 'hs-languagepreference: en-US' -H 'hs-client-version: 301040000' -H 'hs-client-platform: iOS' -H 'Accept-Language: en-us' -H 'Authorization: Bearer {super_long_bearer_token}' -H 'Accept-Encoding: br, gzip, deflate' -H 'tags: ' -H 'Cookie:' > headspace.json
```

Make sure your shit worked by opening `headspace.json` and seeing that there's contents. If not, maybe your token was malformed, or, this issue is out of date. 

Now, create a python file with the following script in it, in the same directory as `headspace.json`:

```
import json
import requests
import sys
import re
import os

token = sys.argv[1] if len(sys.argv) > 1 else None

if not token:
    print("Error: Bearer token required. Grab your token by logging into the Headspace browser app on a computer and provide it to this program.")
    sys.exit(1)

headspace_json = 'headspace.json'
headers = {
    'Accept': '*/*',
    'hs-languagepreference': 'en-US',
    'hs-client-version': '301030000',
    'hs-client-platform': 'iOS',
    'Accept-Language': 'en-us',
    'Authorization': f"Bearer {token}",
    'Accept-Encoding': 'br, gzip, deflate',
    'tags': '',
    'Cookie': ''
}


def safe_filename(filename):
    # Replace reserved characters with an underscore
    safe_name = "".join([c for c in filename if re.match(r'\w', c)])
    return safe_name


def get_sleepcast_info_url(sleepcast_item):
    entity_id = sleepcast_item['attributes']['entityId']
    url = 'https://api.prod.headspace.com/content/sleepcasts/%(entity_id)s' % {
      "entity_id": entity_id
    }
    return url


def fetch_sleepcast_info(sleepcast_item):
    url = get_sleepcast_info_url(sleepcast_item)
    response = requests.get(url, headers=headers, verify=False)
    sleepcast_info = response.json()
    return sleepcast_info


def sleepcast_download_url(content_id):
    return f"https://api.prod.headspace.com/content-aggregation/v1/content/media-items/{content_id}/download?container=aac&tag="


def fetch_write_audio(filename, url):
    audio_raw = requests.get(url, headers=headers, verify=False)
    with open(f"{filename}", 'wb') as f:
        f.write(audio_raw.content)


def get_voice_and_ambience(sleepcast_info):
    ambience_id = sleepcast_info['data']['attributes']['dailySession']['primaryMediaId']
    narration_id = sleepcast_info['data']['attributes']['dailySession']['secondaryMediaId']
    sleepcast_name = sleepcast_info['data']['attributes']['name']
    ambience_path = f"out/{safe_filename(sleepcast_name)}_ambience.mp4"
    narration_path = f"out/{safe_filename(sleepcast_name)}_narration.mp4"
    if not os.path.isfile(ambience_path):
        fetch_write_audio(ambience_path, sleepcast_download_url(ambience_id))
    if not os.path.isfile(narration_path):
        fetch_write_audio(narration_path, sleepcast_download_url(narration_id))


def download_sleepcasts():
    # type library-content-tile
    # Open the JSON file for reading
    with open(headspace_json, 'r') as file:
        # Parse the JSON file data into a Python object
        sleepcast_items = json.load(file)['included']
        for sleepcast in sleepcast_items:
            if sleepcast['type'] == "library-content-tile":
                info = fetch_sleepcast_info(sleepcast)
                get_voice_and_ambience(info)


download_sleepcasts()
```

Make the python file executable `chmod u+x whateveryounamedit.py`

Create a directory `out/` next to the python file.

Launch it, passing in the bearer token (WITHOUT "Bearer") as an argument:

`./whateveryounamedit.py somesuperlongbearertoken`

It should fetch and download the ambience and narration for whatever sleepcasts were in your json.

If you only have a free Headspace account, I believe you'll get authentication errors when trying to download the non-free sleepcasts. Sorry :( The sleepcasts are really high quality and personally helped me solve my sleep issues, however I disagree with Headspace's locking them down to online-only, and so I wanted them offline for when I'm (frequently) without internet access. If you pay them, I believe ethically you should be allowed to access the files offline.

**Comment by doz10us** [comment]
@komali2 Would the code work with courses as well?
I imagine that requires an adjustment in `get_sleepcast_info_url()`.

**Comment by komali2** [comment]
@doz10us yeah basically you'd need to find the requisite URL from the `api` file at the root of this repo and probably change a couple places

**Comment by doz10us** [comment]
@komali2 Nah, they threatened the author of original repo. They requested removal of all code.
I forked it along with thousands others and Haier had to swallow it and back away.

Anyway, that's also why I'm interested in your script: I pay for Headspace subscription for a few years already, and likely will. But constant changes in app annoy me. I already paid for the meditations, now let me access them without clicking a dozen app screens.

**Comment by Joexv** [comment]
That looks great! This is due for a rewrite anyways, the source is fairly messy but Im glad someone else is tackling this issue too!

**Comment by komali2** [comment]
Yeah re: messiness sorry about that, I threw it together in a brief 2 hours slot of free time I never get.

@doz10us I recommend the "rollify" app as the only android app I found that lets you mix arbitrary audio files at various volumes, to replicate the behavior of proprietary white noise and narration apps like White Noise or Headspace. There's some bugginess right now with imported files that I'm fixing to chat with the dev about soon but in my searches it's literally the only app that allows this.

**Comment by j3nesis87** [comment]
Unfortunately not working for me (Sept '24), the headspace.json gives this error (yes I have a paid Headspace account):

{"errors":[{"status":401,"title":"Unauthorized","detail":"{\"title\":\"Unauthorized\",\"status\":401,\"detail\":\"Invalid credentials\",\"data\":null}","data":null}]}

**Comment by sporkbyspork** [comment]
I also get the same 401 error with the longer token from the Request headers.

**Comment by j3nesis87** [comment]
I am still getting the same error. I have tried from both the Respone and Request headers, with Raw mode on.  

I will assume I'm doing something wrong, so forgive the rubber-duck approach here... I'm also fairly new to Linux as my OS so...

I used the curl script in the OP. I have a headspace.json file, that currently says 
`{"errors":[{"status":401,"title":"Unauthorized","detail":"{\"title\":\"Unauthorized\",\"status\":401,\"detail\":\"Invalid credentials\",\"data\":null}","data":null}]}`

I made 'headspace.py' in the same directory, and copied the code from the OP. 

I made a directory simply called `out` in the same directory.

I made an executable file, 'downloadsleepcasts', that goes:
`./headspace.py <bearercode>`

If I execute 'downloadsleepcasts', a terminal flashes, my system hangs for a moment. If I click somewhere, the terminal disappears, nothing appears in /out. However these files have appeared in the ```
directory:
json
os
re
requests 
sys
```

**Comment by DanielBoxer** [comment]
@j3nesis87 Your process looks fine, I think the issue is with your token if you get the 401 error in your json file.

Maybe you're double clicking the token before copying it? If you double click, the selection will stop if there's a period in the token and you won't get the full one.

These are the exact steps I used to get the token, hopefully this works for you:

1. Log into headspace on website
2. Open network tab in inspect mode
3. Reload page
4. Click the first one with domain of `api.prod.headspace.com`
5. Find the request headers
6. Turn raw mode on
7. Triple click the line starting with `Authorization`
8. Copy the selection
9. In notepad, cut the `Authorization: Bearer ` part
10. Copy the remaining token
11. Replace `{super_long_bearer_token}` with the token in the curl command

**Comment by j3nesis87** [comment]
> @j3nesis87 Your process looks fine, I think the issue is with your token if you get the 401 error in your json file.
> 
> Maybe you're double clicking the token before copying it? If you double click, the selection will stop if there's a period in the token and you won't get the full one.
> 
> These are the exact steps I used to get the token, hopefully this works for you:
> 
>     1. Log into headspace on website
> 
>     2. Open network tab in inspect mode
> 
>     3. Reload page
> 
>     4. Click the first one with domain of `api.prod.headspace.com`
> 
>     5. Find the request headers
> 
>     6. Turn raw mode on
> 
>     7. Triple click the line starting with `Authorization`
> 
>     8. Copy the selection
> 
>     9. In notepad, cut the `Authorization: Bearer ` part
> 
>     10. Copy the remaining token
> 
>     11. Replace `{super_long_bearer_token}` with the token in the curl command

I am an absolute numpty. Dunce cap time. Not only had I not been copying the whole code (**triple click works!**), I hadn't been putting the bearer token in the curl command! I had only been putting it in headspace.py.  

However....... it is still not working. 

headspace.json now has content, so that part must be working. Therefore, my `downloadsleepcasts` must be wrong. I have...

`./headspace.py eyJhb.......WggA`

With the rest of the token in ...., naturally. 

Upon executing it, same result - it hangs for a bit, terminal appears then disappears, some random small files appear, and nothing appears in the `out` folder..... feel like I'm doing something really daft (still!) :)

**Comment by mwdossantos** [comment]
For anyone interested I have translated the script to work as a Node.js app, locally, with a simple UI (for the bearer token and headspace.json file) - let me know if you're interested and I can post the workings here. Might be easier to some!

**Comment by DanielBoxer** [comment]
@j3nesis87 You don't need the `downloadsleepcasts` file. Just run it like this in the terminal: `python headspace.py <bearercode>`. You'll probably get some warnings about certificates but you can ignore those.

Also these are some scripts I used after downloading the sleepcasts that you could use if you want to play them using the Rolify app which needs mp3 files.

1. Sorting the out directory

```python
import os
import shutil

out_folder = "out"

files = os.listdir(out_folder)
file_pairs = {}

for file in files:
    if "_ambience" in file or "_narration" in file:
        common_prefix = file.split("_")[0]
        if common_prefix not in file_pairs:
            file_pairs[common_prefix] = []
        file_pairs[common_prefix].append(file)

for common_prefix, file_list in file_pairs.items():
    folder_path = os.path.join(out_folder, common_prefix)
    os.makedirs(folder_path, exist_ok=True)

    for file in file_list:
        source = os.path.join(out_folder, file)
        destination = os.path.join(folder_path, file)
        shutil.move(source, destination)

    print(f"Moved {file_list} into {folder_path}")

print("File sorting complete!")
```

2. Converting from mp4 to mp3 (this takes a long time)

`pip install pydub` and `sudo apt install ffmpeg`

```python
import os

from pydub import AudioSegment

out_folder = "out"

for root, dirs, files in os.walk(out_folder):
    for file in files:
        if file.lower().endswith(".mp4"):
            file_path = os.path.join(root, file)

            mp3_folder = os.path.join(root, "mp3")
            os.makedirs(mp3_folder, exist_ok=True)

            mp3_file_name = os.path.splitext(file)[0] + ".mp3"
            mp3_file_path = os.path.join(mp3_folder, mp3_file_name)

            print(f"Converting {file_path} to {mp3_file_path}")

            try:
                audio = AudioSegment.from_file(file_path, format="mp4")
                audio.export(mp3_file_path, format="mp3")
            except Exception as e:
                print(f"Error converting {file_path}: {e}")

print("All .mp4 files have been converted to .mp3.")
```

**Comment by j3nesis87** [comment]
@DanielBoxer  Brilliant, that's what I was missing, thank you! The reason for me having a file to run is I'm a simple soul who likes to click on things to run, rather than having to type it all the time. But now I know how to make it work :) thank you.

**Comment by jcme** [comment]
Thanks for this!

I put together a quick bash script to run through all of the sorted and converted mp3 files, to merge together ambience and narration for each, and then move each combined file into a "MergedTracks" folder.

Create ```merge_tracks.sh``` in your base folder (where the python scripts have gone).

Add the following contents to the file:

```
#!/bin/bash

BASE_DIR="out"
OUTPUT_DIR="$BASE_DIR/MergedTracks"

mkdir -p "$OUTPUT_DIR"

TRACK_NUMBER=1

for FOLDER in "$BASE_DIR"/*; do
    if [ -d "$FOLDER" ]; then
        TRACK_NAME=$(basename "$FOLDER")
        
        NARRATION_FILE=$(find "$FOLDER" -type f -name "*_narration.mp3")
        AMBIENCE_FILE=$(find "$FOLDER" -type f -name "*_ambience.mp3")
        
        if [ -f "$NARRATION_FILE" ] && [ -f "$AMBIENCE_FILE" ]; then
            TRACK_NUMBER_PADDED=$(printf "%02d" "$TRACK_NUMBER")
            OUTPUT_FILE="$OUTPUT_DIR/$TRACK_NUMBER_PADDED $TRACK_NAME.mp3"
            
            ffmpeg -i "$NARRATION_FILE" -i "$AMBIENCE_FILE" -filter_complex amix=inputs=2:duration=longest "$OUTPUT_FILE" -y
            
            TRACK_NUMBER=$((TRACK_NUMBER + 1))
        else
            echo "Warning: Missing narration or ambience file in $FOLDER"
        fi
    fi
done
```

Make it executable: ```chmod +x merge_tracks.sh```
Then run it with ```./merge_tracks.sh```

It'll take some time, but once it's done you should have all the merged mp3s neatly arranged in ```out/MergedTracks```, ready for copying to a media server of choice

**Comment by komali2** [comment]
That's very cool :) I like them separate because I prefer being able to control the volume of each independently

**Comment by chrisadesign** [comment]
> For anyone interested I have translated the script to work as a Node.js app, locally, with a simple UI (for the bearer token and headspace.json file) that also combines the files automatically after download (while preserving the original narration and ambience files) - let me know if you're interested and I can post the workings here. Might be easier to some!

Hey @mwdossantos can you post this?

**Comment by jasperhooloop** [comment]
> For anyone interested I have translated the script to work as a Node.js app, locally, with a simple UI (for the bearer token and headspace.json file) that also combines the files automatically after download (while preserving the original narration and ambience files) - let me know if you're interested and I can post the workings here. Might be easier to some!

Would loooove that! Come back lol
